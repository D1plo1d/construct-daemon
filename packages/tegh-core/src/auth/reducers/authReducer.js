import { loop, Cmd } from 'redux-loop'
import { Map, Record } from 'immutable'
import { ec as EC } from 'elliptic'

import displayInviteInConsole from '../sideEffects/displayInviteInConsole'

import { SET_CONFIG } from '../../config/actions/setConfig'

import createInvite from '../actions/createInvite'
import { DAT_PEER_HANDSHAKE_RECEIVED } from '../actions/datPeerHandshakeReceived'
import { DAT_PEER_DATA_RECEIVED } from '../actions/datPeerDataReceived'

const ec = new EC('curve25519')

export const initialState = Record({
  hostKeys: null,
  usersAndInvitesByPublicKey: Map(),
  datSessionsByPeerID: Map(),
})

const authReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CONFIG: {
      const { config } = action.payload
      const { users, invites } = config.auth

      const admins = users.filter(user => user.admin)
      const adminInvites = invites.filter(invite => invite.admin)
      let consoleInvites = invites.filter(invite => invite.displayInConsole)

      const sideEffects = []

      /*
       * If no invites exist and no admins exist then create an invite so the
       * initial user can connect
       */
      if (admins.size === 0 && adminInvites.size === 0) {
        const nextAction = createInvite({
          displayInConsole: true,
          admin: true,
        })
        consoleInvites = consoleInvites.push(nextAction.payload.invite)
        sideEffects.push(Cmd.action(nextAction))
      }

      /*
       * Display invites in the console
       */
      consoleInvites.forEach(invite => (
        sideEffects.push(Cmd.run(displayInviteInConsole, {
          args: [{ hostPublicKey: hostKeys.getPublic(), invite }],
        }))
      ))

      return loop(state, Cmd.list(sideEffects))
    }
    case DAT_PEER_HANDSHAKE_RECEIVED: {
      const hostEphemeralKeys = ec.genKeyPair()
      hostEphemeralKeys.getPublic()
    }
    case DAT_PEER_DATA_RECEIVED: {
    }
    default: {
      return state
    }
  }
}

export default authReducer
