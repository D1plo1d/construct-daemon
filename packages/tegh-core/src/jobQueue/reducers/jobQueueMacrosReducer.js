import { loop, Cmd } from 'redux-loop'

/* config actions */
import requestSpoolJobFile from '../../spool/actions/requestSpoolJobFile'
import requestSpoolNextJobFile from '../../spool/actions/requestSpoolNextJobFile'
import despoolCompleted from '../../spool/actions/despoolCompleted'
import { DESPOOL_TASK } from '../../spool/actions/despoolTask'

const jobQueueMacrosReducer = (state, action) => {
  switch (action.type) {
    case DESPOOL_TASK: {
      const { macro, args } = action.payload
      let nextAction = null

      switch (macro) {
        case 'spoolNextJobFile': {
          nextAction = requestSpoolNextJobFile()
          break
        }
        case 'spoolJobFile': {
          nextAction = requestSpoolJobFile(args)
          break
        }
        default: {
          return
        }
      }

      return loop(
        state,
        Cmd.list([
          Cmd.action(nextAction),
          Cmd.action(despoolCompleted()),
        ]),
      )
    }
    default: {
      return state
    }
  }
}

export default jobQueueMacrosReducer
