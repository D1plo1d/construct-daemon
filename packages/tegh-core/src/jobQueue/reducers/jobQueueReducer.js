import { Record, Map, List } from 'immutable'
import { loop, Cmd } from 'redux-loop'

import unlinkTmpFiles from '../sideEffects/unlinkTmpFiles'

import JobHistoryEvent from '../types/JobHistoryEvent'

import {
  SPOOL_PRINT,
  START_PRINT,
  CANCEL_PRINT,
  PRINT_ERROR,
  FINISH_PRINT,
} from '../types/JobHistoryTypeEnum'

import getJobTmpFiles from '../selectors/getJobTmpFiles'
import getSpooledJobFiles from '../selectors/getSpooledJobFiles'
import getCompletedJobs from '../selectors/getCompletedJobs'
import getTaskIDByJobFileID from '../selectors/getTaskIDByJobFileID'

import { CREATE_JOB } from '../actions/createJob'
import deleteJob, { DELETE_JOB } from '../actions/deleteJob'

import { SPOOL_TASK } from '../../spool/actions/spoolTask'
import { DESPOOL_TASK } from '../../spool/actions/despoolTask'
import { CANCEL_TASK } from '../../spool/actions/cancelTask'

import { PRINTER_READY } from '../../printer/actions/printerReady'
import { ESTOP } from '../../printer/actions/estop'
import { DRIVER_ERROR } from '../../printer/actions/driverError'

/* reducer */

export const initialState = Record({
  jobs: Map(),
  jobFiles: Map(),
  /*
   * A list of JobFileHistory records which can be reduced to determine the
   * current status of any job or jobFile in the queue.
   */
  history: List(),
})()

const jobQueue = (state = initialState, action) => {
  switch (action.type) {
    case CREATE_JOB: {
      const { job, jobFiles } = action.payload

      return state
        .setIn(['jobs', job.id], job)
        .mergeIn(['jobFiles'], jobFiles)
    }
    case DELETE_JOB: {
      const { jobID } = action.payload

      const tmpFilePaths = getJobTmpFiles(state)({ jobID }).toArray()

      const nextState = state
        .deleteIn(['jobs', jobID])
        .updateIn(['jobFiles'], jobFiles => (
          jobFiles.filter(file => file.jobID !== jobID)
        ))
        .updateIn(['history'], history => (
          history.filter(historyEvent => historyEvent.jobID !== jobID)
        ))

      return loop(
        nextState,
        Cmd.run(unlinkTmpFiles, { args: tmpFilePaths }),
      )
    }
    case PRINTER_READY:
    case ESTOP:
    case DRIVER_ERROR: {
      const isCancelledByUser = action.type === ESTOP
      const eventType = isCancelledByUser ? CANCEL_PRINT : PRINT_ERROR

      const taskIDs = getTaskIDByJobFileID(state)

      /* error or cancel any printing job file */
      const events = getSpooledJobFiles(state).map(jobFile => (
        JobHistoryEvent({
          jobID: jobFile.jobID,
          jobFileID: jobFile.id,
          taskID: taskIDs.get(jobFile.id),
          type: eventType,
        })
      ))
      return state.update('history', history => history.concat(events))
    }
    case CANCEL_TASK: {
      const { taskID } = action.payload
      const jobFileID = getTaskIDByJobFileID(state).findKey(v => v === taskID)

      if (jobFileID == null) return state

      const { jobID } = state.jobFiles.get(jobFileID)

      const historyEvent = JobHistoryEvent({
        jobID,
        jobFileID,
        taskID,
        type: CANCEL_PRINT,
      })

      /* mark each spooled job file as cancelled */
      return state.update('history', history => history.push(historyEvent))
    }
    case SPOOL_TASK: {
      const {
        jobID,
        jobFileID,
        id: taskID,
      } = action.payload.task

      if (jobID == null) return state

      /*
       * record the spooling of the print in the job history
       */
      const historyEvent = JobHistoryEvent({
        jobID,
        jobFileID,
        taskID,
        type: SPOOL_PRINT,
      })

      const nextState = state
        .update('history', history => history.push(historyEvent))

      /*
       * delete the previous job upon spooling a subsequent job
       */
      const jobsForDeletion = getCompletedJobs(state).toList()

      if (jobsForDeletion.size > 1) {
        throw new Error('only one completed Job should exist at a time')
      }

      if (jobsForDeletion.size === 1) {
        const nextAction = deleteJob({ jobID: jobsForDeletion.get(0).id })

        return loop(nextState, Cmd.action(nextAction))
      }

      return nextState
    }
    case DESPOOL_TASK: {
      const {
        jobID,
        jobFileID,
        currentLineNumber,
        data,
      } = action.payload.task

      if (jobID == null) return state

      const eventType = (() => {
        if (currentLineNumber === 0) return START_PRINT
        if (currentLineNumber === data.size - 1) return FINISH_PRINT
        return null
      })()

      if (eventType == null) return state

      const taskID = getTaskIDByJobFileID(state).get(jobFileID)

      /*
       * record the start or finish of the print in the job history
       */
      const historyEvent = JobHistoryEvent({
        jobID,
        jobFileID,
        taskID,
        type: eventType,
      })

      return state.update('history', history => history.push(historyEvent))
    }
    default: {
      return state
    }
  }
}

export default jobQueue
