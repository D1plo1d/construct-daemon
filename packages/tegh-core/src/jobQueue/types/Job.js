// @flow
import uuid from 'uuid/v4'
import { Record } from 'immutable'

import type { RecordOf } from 'immutable'

export type JobT = RecordOf<{
  id: string,
  name: string,
  createdAt: ?number,
  quantity: number,
}>

const JobRecord = Record({
  id: null,
  name: null,
  createdAt: null,
  quantity: 1,
})

const Job = attrs => JobRecord({
  id: uuid(),
  createdAt: new Date().toISOString(),
  ...attrs,
})

export const MockJob = attrs => Job({
  name: 'test.ngc',
  ...attrs,
})


export default Job
