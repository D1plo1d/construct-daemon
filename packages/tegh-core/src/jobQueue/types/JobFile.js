// @flow
import uuid from 'uuid/v4'
import { Record } from 'immutable'

import type { RecordOf } from 'immutable'

import { SPOOLED } from './JobStatusEnum'

export type JobFileT = RecordOf<{
  name: string,
  filePath: string,
  isTmpFile: boolean,
  quantity: number,
  jobID: string,
}>

const JobFileRecord = Record({
  id: null,
  name: null,
  filePath: null,
  isTmpFile: null,
  quantity: 1,
  status: SPOOLED,
  tasksCompleted: 0,
  jobID: null,
})

const JobFile = attrs => JobFileRecord({
  id: uuid(),
  ...attrs,
})

export const MockJobFile = JobFile

export default JobFile
