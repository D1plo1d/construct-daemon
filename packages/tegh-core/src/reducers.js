export config from './config/reducers/configReducer'
export devices from './devices/reducers/devicesReducer'
export jobQueue from './jobQueue/reducers/jobQueueReducer'
export log from './log/reducers/logReducer'
export macros from './macros/reducers/macrosReducer'
export pluginManager from './pluginManager/reducers/pluginManagerReducer'
export spool from './spool/reducers/spoolReducer'
export status from './printer/reducers/statusReducer'
