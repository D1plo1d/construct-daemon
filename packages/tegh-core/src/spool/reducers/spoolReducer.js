import { loop, Cmd } from 'redux-loop'
import {
  Record, List, Map,
} from 'immutable'

import isIdle from '../selectors/isIdle'

import {
  EMERGENCY,
  NORMAL,
  PREEMPTIVE,
  priorityOrder,
} from '../types/PriorityEnum'

import {
  PRINTING,
  isSpooled,
} from '../types/TaskStatusEnum'

/* printer actions */
import { PRINTER_READY } from '../../printer/actions/printerReady'
import { ESTOP } from '../../printer/actions/estop'
import { DRIVER_ERROR } from '../../printer/actions/driverError'
/* job actions */
import { CANCEL_JOB } from '../../jobQueue/actions/cancelJob'
import { DELETE_JOB } from '../../jobQueue/actions/deleteJob'
/* task actions */
import { SPOOL_TASK } from '../actions/spoolTask'
import requestDespool, { REQUEST_DESPOOL } from '../actions/requestDespool'
import despoolTask from '../actions/despoolTask'
import { CANCEL_TASK } from '../actions/cancelTask'
import cancelAllTasks, { CANCEL_ALL_TASKS } from '../actions/cancelAllTasks'

export const initialState = Record({
  priorityQueues: Record({
    [EMERGENCY]: List(),
    [PREEMPTIVE]: List(),
    [NORMAL]: List(),
  })(),
  tasks: Map(),
  currentTaskID: null,
})()

const removeTaskReferences = (state) => {
  const nextTaskIDs = state.tasks.map(task => task.id)

  let nextState = state
    .update('priorityQueues', queues => queues.map(queue => (
      queue.filter(taskID => nextTaskIDs.includes(taskID))
    )))

  if (nextTaskIDs.includes(state.currentTaskID) === false) {
    nextState = nextState.set('currentTaskID', null)
  }

  return nextState
}

const spoolReducer = () => (state = initialState, action) => {
  switch (action.type) {
    /* Spool reset actions */
    case PRINTER_READY:
    case ESTOP:
    case DRIVER_ERROR:
    case CANCEL_ALL_TASKS: {
      return initialState
    }
    case CANCEL_JOB:
    case DELETE_JOB: {
      const { jobID } = action.payload

      const nextState = state.update('tasks', tasks => (
        tasks.filter(task => task.jobID !== jobID)
      ))

      return removeTaskReferences(nextState)
    }
    case CANCEL_TASK: {
      const { taskID } = action.payload

      const nextState = state.update('tasks', tasks => tasks.remove(taskID))

      return removeTaskReferences(nextState)
    }
    case SPOOL_TASK: {
      const { payload } = action
      const { id, priority } = payload.task

      let nextState = state

      /*
      * if the task is an emergency then cancel all other tasks in the queue
      */
      if (priority === EMERGENCY) {
        nextState = initialState
      }

      nextState = state.setIn(['tasks', id], payload.task)

      if (
        isIdle.resultFunc(state.tasks) === false
        && priority !== EMERGENCY
        && payload.task.internal !== true
      ) {
        throw new Error('Cannot spool non-emergency tasks when printing a job')
      }


      /* create the task */
      const createAction = createTask({ task: payload.task })
      nextState = taskMap.createOne(nextState, createAction)

      const taskQueue = ['priorityQueues', priority]
      const shouldDespool = nextState.currentTaskID == null

      nextState = nextState
        .updateIn(taskQueue, list => list.push(id))

      /*
       * despool the first line if nothing is spooled
       */
      if (shouldDespool) {
        return loop(nextState, Cmd.action(requestDespool()))
      }

      return nextState
    }
    case REQUEST_DESPOOL: {
      const { priorityQueues, currentTaskID } = state
      let nextState = state
      if (currentTaskID != null) {
        /*
         * despool the next line or finish the task via the taskReducer
         */
        nextState = taskMap.updateOne(nextState, action, currentTaskID)
      }
      const currentTask = nextState.tasks.get(currentTaskID)
      if (currentTask == null || currentTask.status !== PRINTING) {
        /*
         * if the current task is done then despool the next task if there is
         * anything to despool.
         */
        const priority = priorityOrder.find(priorityOption => (
          priorityQueues[priorityOption].size > 0
        ))

        if (priority == null) {
          nextState = nextState.set('currentTaskID', null)
        } else {
          const nextTaskID = state.priorityQueues[priority].first()
          /*
           * start the task via the taskReducer
           */
          nextState = nextState
            .mergeIn(['tasks', nextTaskID], {
              startedAt: new Date().toISOString(),
              status: PRINTING,
              currentLineNumber: 0,
            })
            .set('currentTaskID', nextTaskID)
            .updateIn(['priorityQueues', priority], list => list.shift())
        }
      }

      if (nextState.currentTaskID == null) {
        return nextState
      }

      const nextTask = nextState.tasks.get(nextState.currentTaskID)
      return loop(nextState, Cmd.action(despoolTask(nextTask)))
    }
    case REQUEST_DESPOOL: {
      if (state.currentLineNumber < state.data.size - 1) {
        /*
         * if the task has more lines to execute then increment the line number
         */
        return state.update('currentLineNumber', i => i + 1)
      }
      /* mark tasks as done after they are completed */
      return state.merge({
        // TODO: stoppedAt should eventually be changed to be sent after
        // the printer sends 'ok' or 'error' and should be based off
        // estimated print time
        stoppedAt: new Date().toISOString(),
        status: DONE,
        /* the data of each completed task is deleted to save space */
        data: null,
      })
    }
    default: {
      return state
    }
  }
}

export default spoolReducer
