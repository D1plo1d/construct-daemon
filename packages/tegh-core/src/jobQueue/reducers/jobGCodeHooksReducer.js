import { Record } from 'immutable'
import { loop, Cmd } from 'redux-loop'

import { SET_CONFIG } from '../../config/actions/setConfig'
import spoolTask, { SPOOL_TASK } from '../../spool/actions/spoolTask'
import { DESPOOL_TASK } from '../../spool/actions/despoolTask'
import { NORMAL } from '../../spool/types/PriorityEnum'

import getPluginModels from '../../config/selectors/getPluginModels'

const initialState = Record({
  beforePrintHook: null,
  afterPrintHook: null,
})()

const jobQueueMacrosReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CONFIG: {
      const { config } = action.payload
      const {
        beforePrintHook,
        afterPrintHook,
      } = getPluginModels(config).get('@tegh/core')

      const splitLines = hook => (hook == null ? null : hook.split('\n'))

      return state.merge({
        beforePrintHook: splitLines(beforePrintHook),
        afterPrintHook: splitLines(afterPrintHook),
      })
    }
    case SPOOL_TASK: {
      const { beforePrintHook } = state
      const { jobID } = action.payload.task

      if (jobID == null || beforePrintHook == null) {
        return state
      }

      const taskAttrs = {
        priority: NORMAL,
        internal: true,
        name: '[BEFORE_PRINT_HOOK]',
        data: beforePrintHook,
      }

      return loop(
        state,
        Cmd.action(spoolTask(taskAttrs, { prepend: true })),
      )
    }
    case DESPOOL_TASK: {
      const { afterPrintHook } = state
      const { jobID } = action.payload.task

      if (jobID == null || afterPrintHook == null) {
        return state
      }

      const taskAttrs = {
        priority: NORMAL,
        internal: true,
        name: '[AFTER_PRINT_HOOK]',
        data: afterPrintHook,
      }

      return loop(
        state,
        Cmd.action(spoolTask(taskAttrs, { prepend: true })),
      )
    }
    default: {
      return state
    }
  }
}

export default jobQueueMacrosReducer
