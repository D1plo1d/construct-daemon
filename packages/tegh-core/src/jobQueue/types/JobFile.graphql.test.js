import { Map } from 'immutable'

import snapshotTestGraphQLType from '../../util/testing/snapshotTestGraphQLType'

import { initialState as jobQueueState } from '../reducers/jobQueueReducer'
import { initialState as spoolState } from '../../spool/reducers/spoolReducer'

import Job from './Job'
import JobFile from './JobFile'
// import { NORMAL } from './PriorityEnum'

import JobFileGraphQL from './JobFile.graphql'

const jobID = 'test_job_id'

const job = Job({
  id: jobID,
  name: 'test job',
  internal: false,
  // priority: NORMAL,
  data: ['g1 x10', 'g1 y20'],
})

const jobFile = JobFile({
  id: 'test_job_file_id',
  name: 'gear.gcode',
  quantity: 1,
  jobID,
})

const task = {
  id: '123',
  jobID,
  jobFileID: jobFile.id,
}

const state = {
  jobQueue: jobQueueState.merge({
    jobs: Map({
      [job.id]: job,
    }),
    jobFiles: Map({
      [jobFile.id]: jobFile,
    }),
  }),
  spool: spoolState.merge({
    tasks: Map({
      [task.id]: task,
    }),
  }),
}

snapshotTestGraphQLType('JobFileGraphQL', {
  type: JobFileGraphQL,
  fieldConfigs: {
    tasks: {
      args: { excludeCompletedTasks: false },
    },
  },
  rootValue: jobFile,
  contextValue: {
    store: {
      getState: () => state,
    },
  },
})
