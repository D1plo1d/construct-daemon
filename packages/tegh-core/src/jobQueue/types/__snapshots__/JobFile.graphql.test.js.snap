// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobFileGraphQL presents a consistent API 1`] = `
Object {
  "description": null,
  "enumValues": null,
  "fields": Array [
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "id",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "ID",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "name",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "String",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "quantity",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [
        Object {
          "defaultValue": null,
          "description": null,
          "name": "excludeCompletedTasks",
          "type": Object {
            "kind": "SCALAR",
            "name": "Boolean",
            "ofType": null,
          },
        },
      ],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "tasks",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "LIST",
          "name": null,
          "ofType": Object {
            "kind": "OBJECT",
            "name": "Task",
            "ofType": null,
          },
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "tasksCompleted",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "totalTasks",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "SCALAR",
          "name": "Int",
          "ofType": null,
        },
      },
    },
    Object {
      "args": Array [],
      "deprecationReason": null,
      "description": null,
      "isDeprecated": false,
      "name": "status",
      "type": Object {
        "kind": "NON_NULL",
        "name": null,
        "ofType": Object {
          "kind": "ENUM",
          "name": "JobStatus",
          "ofType": null,
        },
      },
    },
  ],
  "inputFields": null,
  "interfaces": Array [],
  "kind": "OBJECT",
  "name": "JobFile",
  "possibleTypes": null,
}
`;

exports[`JobFileGraphQL resolves every field without error 1`] = `
Object {
  "id": "test_job_file_id",
  "name": "gear.gcode",
  "quantity": 1,
  "status": "QUEUED",
  "tasks": Array [
    Object {
      "id": "[DYNAMIC FIELD OMITTED]",
    },
  ],
  "tasksCompleted": 0,
  "totalTasks": 1,
}
`;
